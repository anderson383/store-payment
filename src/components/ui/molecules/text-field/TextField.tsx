import styles from './TextField.module.scss';
import { useField } from 'formik';
import { useEffect } from 'react';
interface TextFieldProps {
  name: string;
  placeholder?: string;
  type?: string;
  className?: string;
  label?: string;
  disabled?: boolean;
  disableError?: boolean;
  props?: Record<string, any>;
  variant?: 'filled' | 'outlined' | 'standard';
  size?: 'fullwidth' | 'md';
}

export const TextField:React.FC<TextFieldProps> = ({
  name,
  placeholder,
  type = 'text',
  className,
  label,
  disableError = false,
  props,
  variant = 'filled',
  disabled,
  size = 'md'
}) => {
  const [field, meta, helpers] = useField(name);
  
  useEffect(() => {
    props?.onChange && props.onChange(field.value)
  }, [field.value])

  return (
    <div className={`${styles.textField} ${styles['textField__' + size]}`}>
      <label htmlFor={name}> { label } </label>
      <input
        id={name}
        disabled={disabled}
        placeholder={placeholder}
        className={className}
        {...props}
        {...field}
        type={type}
      />
      {
       (meta?.error && meta.touched) ? <span className={styles.error}> {meta.error} </span> : null
      }
    </div>
  );
};
