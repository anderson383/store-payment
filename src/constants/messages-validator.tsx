
export const creditCardRequired = 'El número de tarjeta es obligatorio';

export const creditCardMaxDigit = 'El número de tarjeta debe tener 16 dígitos';

export const cardHolderNameRequired = 'El nombre es obligatorio';

export const expiresRequired = 'La fecha de expiración es obligatoria';

export const cvcRequired = 'El CVC es obligatorio';

export const cvcMaxDigit = 'El CVC debe tener 3 o 4 dígitos';
