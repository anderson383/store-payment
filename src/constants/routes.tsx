import { IndexPage } from '../pages/index';
import { DetailPage } from '../pages/detail/detail';

export const ROUTES:Record<string, {path: string, component: React.ReactNode}> = {
  index: {
    path: '/',
    component: <IndexPage />
  },
  detailproduct: {
    path: '/detail/:id',
    component: <DetailPage />
  }
}