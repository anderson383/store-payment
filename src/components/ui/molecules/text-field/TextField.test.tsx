import * as Yup from 'yup';

import {
  fireEvent,
  render,
  screen
} from '@testing-library/react';
import { act } from 'react';
import { TextField } from './TextField';
import { Formik } from 'formik';
import { fieldRequired } from '../../../../constants/messages-validator';

const onChangeInputMock = jest.fn()

const validationSchemaText =  Yup.object({
  test: Yup.string().required(fieldRequired)
})


describe('Tests in TextField component', () => {

  const setup = (ui: React.ReactElement, { initialValues, ...renderOptions }: any) => {
    return render(
      <Formik initialValues={initialValues} onSubmit={() => {}} validationSchema={validationSchemaText} >
        {ui}
      </Formik>,
      renderOptions
    );
  };

  test('Should render the component', async () => {
    const {asFragment} = setup(<TextField name="test" />, { initialValues: { test: '' } });
    
    expect(asFragment()).toMatchSnapshot();
  });

  test('Should onBlur and render error', async () => {
    const {asFragment} = setup(<TextField type='text' label='test' name="test" />, { initialValues: { test: '' } });
    const input = screen.getByLabelText(/test/i);
    fireEvent.blur(input);
    fireEvent.change(input, { target: { value: null } });
    await act(() => Promise.resolve())
    expect(screen.getByText(fieldRequired)).toBeDefined();
  });

  test('Should listening changes', async () => {
    const {asFragment} = setup(<TextField props={{ onChange: onChangeInputMock }} label='test' name="test" />, { initialValues: { test: '' } });
    const input = screen.getByLabelText(/test/i);
    fireEvent.blur(input);
    fireEvent.change(input, { target: { value: 'test' } });
    await act(() => Promise.resolve())
    expect(onChangeInputMock).toHaveBeenCalled();
  });
});
