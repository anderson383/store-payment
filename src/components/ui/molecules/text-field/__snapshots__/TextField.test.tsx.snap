// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in TextField component Should render the component 1`] = `
<DocumentFragment>
  <div
    class="textField textField__md"
  >
    <label
      for="test"
    />
    <input
      id="test"
      name="test"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
